go/connection_pool.go

```diff
@@ -0,0 +1,233 @@
+	convAddress, err := DomainToIP(addresses)
+	if err != nil {
+		return fmt.Errorf("Failed to find IP, error: %s ", err.Error())
+	}
+
+	pool.addresses = convAddress
+	pool.conf = conf
+	pool.hostIndex = 0
+	pool.log = log
+
+	// Check config
+	pool.conf.validateConf(pool.log)
+	// Check input
+	if len(addresses) == 0 {
+		return fmt.Errorf("Failed to initialize connection pool: illegal address input")
+	}
+	if &conf == nil {
+		return fmt.Errorf("Failed to initialize connection pool: no configuration")
+	}
+
+	for i := 0; i < pool.conf.MinConnPoolSize; i++ {
+		// Simple round-robin
+		newConn := newConnection(pool.addresses[i%len(addresses)])
+
+		// Open connection to host
+		err := newConn.open(newConn.severAddress, pool.conf.TimeOut)
+		if err != nil {
+			// If initialization failed, clean idle queue
+			idleLen := pool.idleConnectionQueue.Len()
+			for i := 0; i < idleLen; i++ {
+				pool.idleConnectionQueue.Front().Value.(*connection).close()
+				pool.idleConnectionQueue.Remove(pool.idleConnectionQueue.Front())
+			}
+			return fmt.Errorf("Failed to open connection, error: %s ", err.Error())
+		}
+		// Mark connection as in use
+		pool.idleConnectionQueue.PushBack(newConn)
+	}
+	pool.log.Info("connection pool is initialized successfully")
+	return nil
+}
+
+func (pool *ConnectionPool) GetSession(username, password string) (*Session, error) {
+	// Get valid and usable connection
+	var conn *connection = nil
+	var err error = nil
+	const retryTimes = 3
+	for i := 0; i < retryTimes; i++ {
+		conn, err = pool.getIdleConn()
+		if err == nil {
+			break
+		}
+	}
+	if conn == nil {
+		return nil, err
+	}
+	// Authenticate
+	resp, err := conn.authenticate(username, password)
+	if err != nil || resp.GetErrorCode() != graph.ErrorCode_SUCCEEDED {
+		// if authentication failed, put connection back
+		pool.rwLock.Lock()
+		defer pool.rwLock.Unlock()
+		removeFromList(&pool.activeConnectionQueue, conn)
+		pool.idleConnectionQueue.PushBack(conn)
+		return nil, err
+	}
+
+	sessID := resp.GetSessionID()
+	// Create new session
+	newSession := Session{
+		sessionID:  sessID,
+		connection: conn,
+		connPool:   pool,
+		log:        pool.log,
+	}
+
+	return &newSession, nil
+}
+
+func (pool *ConnectionPool) getIdleConn() (*connection, error) {
+	pool.rwLock.Lock()
+	defer pool.rwLock.Unlock()
+
+	// Take an idle valid connection if possible
+	if pool.idleConnectionQueue.Len() > 0 {
+		var newConn *connection = nil
+		var newEle *list.Element = nil
+		for ele := pool.idleConnectionQueue.Front(); ele != nil; ele = ele.Next() {
+			// Check if connection is valid
+			if res := ele.Value.(*connection).ping(); res == true {
+				newConn = ele.Value.(*connection)
+				newEle = ele
+				break
+			}
+		}
+		if newConn == nil {
+			newConn, err := pool.createConnection()
+			return newConn, err
+		}
+		// Remove new connection from idle and add to active if found
+		pool.idleConnectionQueue.Remove(newEle)
+		pool.activeConnectionQueue.PushBack(newConn)
```

Comment:
the `createConnection` has pushback the `newConn`, the `newConn` has input twice.it will cause two thread to use the same connection.